src/tribler/core/components/ipv8/eva_protocol.py

```diff
@@ -503,76 +556,87 @@ def send_scheduled(self):
-        self._notify_error(peer, e)
-
-    def _notify_error(self, peer, exception):
-        logger.warning(f'Exception.Peer hash {hash(peer)}: "{exception}"')
+        self._notify_error(peer, e, future)
 
+    def _notify_error(self, peer, exception, future: Optional[Future] = None):
+        logger.warning(f'Exception.Peer hash {peer}: "{exception}"')
+        if future:
+            future.set_exception(exception)
         for callback in self.error_callbacks:
             callback(peer, exception)
 
-    def _schedule_terminate(self, container, peer, transfer):
-        if not self.terminate_by_timeout_enabled:
-            return
+    async def _terminate_by_timeout_task(self, container, peer, transfer):
+        remaining_time = self.timeout_interval_in_sec
 
-        self.community.register_anonymous_task('eva_terminate_by_timeout', self._terminate_by_timout_task, container,
-                                               peer, transfer, delay=self.timeout_interval_in_sec, )
+        while self.terminate_by_timeout_enabled:
+            await asyncio.sleep(remaining_time)
+            if transfer.released:
+                return
 
-    def _terminate_by_timout_task(self, container, peer, transfer):
-        if transfer.released or not self.terminate_by_timeout_enabled:
-            return
+            remaining_time = self.timeout_interval_in_sec - (time.time() - transfer.updated)
+            if remaining_time <= 0:  # it is time to terminate
+                self.terminate(container, peer, transfer)
+                exception = TimeoutException('Terminated by timeout', transfer)
+                self._notify_error(peer, exception, transfer.future)
+                return
 
-        timeout = self.timeout_interval_in_sec
-        remaining_time = timeout - (time.time() - transfer.updated)
+    async def _resend_acknowledge_task(self, peer, transfer):
+        remaining_time = self.retransmit_interval_in_sec
 
-        if remaining_time > 0:
-            self.community.register_anonymous_task('eva_terminate_by_timeout', self._terminate_by_timout_task,
-                                                   container, peer, transfer, delay=remaining_time, )
-            return
+        while self.retransmit_enabled:
+            await asyncio.sleep(remaining_time)
 
-        EVAProtocol.terminate(container, peer, transfer)
-        self._notify_error(peer, TimeoutException(f'Terminated by timeout. Timeout is: {timeout} sec', transfer))
+            attempts_are_over = transfer.attempt >= self.retransmit_attempt_count
+            if attempts_are_over or transfer.released:
+                return
 
-    def _schedule_resend_acknowledge(self, peer, transfer):
-        if not self.retransmit_enabled:
-            return
+            remaining_time = self.retransmit_interval_in_sec - (time.time() - transfer.updated)
+            if remaining_time <= 0:  # it is time to retransmit
```

Comment:
**nitpick:** Probably the expression can be a tiny bit more clear if comparing not with zero, but with `self.retransmit_interval_in_sec`:
```suggestion
            time_since_last_update = time.time() - transfer.updated
            if time_since_last_update > self.retransmit_interval_in_sec:  # it is time to retransmit
```

Reply:
Your suggestion has been added.

Reply:
Yes, but `remaining_time` is used during the next loop iteration:

```python
        while self.terminate_by_timeout_enabled:
            await asyncio.sleep(remaining_time)
            ...
            remaining_time = self.timeout_interval_in_sec - (time.time() - transfer.updated)
```